game-core/src/main/java/games/strategy/triplea/ai/pro/util/MoveBatcher.java

```diff
@@ -0,0 +1,161 @@
+    private final ArrayList<Unit> units;
+    private final Route route;
+    private final @Nullable HashMap<Unit, Unit> unitsToTransports;
+
+    private Move(
+        final ArrayList<Unit> units,
+        final Route route,
+        final @Nullable HashMap<Unit, Unit> unitsToTransports) {
+      this.units = units;
+      this.route = route;
+      this.unitsToTransports = unitsToTransports;
+    }
+
+    Move(final ArrayList<Unit> units, final Route route) {
+      this(units, route, null);
+    }
+
+    Move(final Unit unit, final Route route, final Unit transportToLoad) {
+      this(new ArrayList<>(List.of(unit)), route, new HashMap<>(Map.of(unit, transportToLoad)));
+    }
+
+    private boolean isTransportLoad() {
+      return this.unitsToTransports != null;
+    }
+
+    boolean canMergeWith(final Move other) {
+      return other.isTransportLoad() == isTransportLoad() && route.equals(other.route);
+    }
+
+    boolean mergeWith(final Move other) {
+      if (canMergeWith(other)) {
+        // Merge units and transports.
+        units.addAll(other.units);
+        if (isTransportLoad()) {
+          unitsToTransports.putAll(other.unitsToTransports);
+        }
+        return true;
+      }
+      return false;
+    }
+
+    void addTo(
+        final List<Collection<Unit>> moveUnits,
+        final List<Route> moveRoutes,
+        final List<Map<Unit, Unit>> unitsToTransports) {
+      moveUnits.add(units);
+      moveRoutes.add(route);
+      unitsToTransports.add(this.unitsToTransports);
+    }
+  }
+
+  private final ArrayList<ArrayList<Move>> moveSequences = new ArrayList<>();
+
+  /**
+   * Starts a new sequence. This must be called before calling any add*() methods. A sequence
+   * indicates a logical dependency relationship between the moves in that sequence.
+   */
+  public void newSequence() {
+    moveSequences.add(new ArrayList<>());
+  }
+
+  /**
+   * Adds a transport load move.
+   *
+   * @param unit The unit to move.
+   * @param route The route to move on.
+   * @param transport The transport to load.
+   */
+  public void addTransportLoad(final Unit unit, final Route route, final Unit transport) {
+    addMove(new Move(unit, route, transport));
+  }
+
+  /**
+   * Adds a move for the specified units. The units list must not be changed after adding the move.
+   *
+   * @param units The units to move. Note that type is ArrayList explicitly to prevent an
+   *     unmodifiable List.of() being passed.
+   * @param route The route to move on.
+   */
+  public void addMove(final ArrayList<Unit> units, final Route route) {
+    addMove(new Move(units, route));
+  }
+
+  private void addMove(final Move newMove) {
+    final ArrayList<Move> sequence = moveSequences.get(moveSequences.size() - 1);
+    if (!sequence.isEmpty() && sequence.get(sequence.size() - 1).mergeWith(newMove)) {
+      return;
+    }
+    sequence.add(newMove);
+  }
+
+  /**
+   * Emits the accumulated moves to the passed output params.
```

Comment:
I think it should be:

Collection<Move> moves = moveBatcher.batchAndEmit();

But this related to the previously discussed refactoring about making IMoveDelegate and ProMoveUtils operate on Move objects instead of three parallel lists.

Doing so would be an even bigger change than this PR. Can we do it in the next PR like agreed? I feel like all your comments are around this - which is an existing bad pattern in the current code that I'm not changing, just making my class work with it.

Like, let's not put effort into polishing this existing ugly mess if the plan is to refactor it. Because that's just wasted effort.

Reply:
Sounds reasonable. I can't read your mind and what you intend for a future change, it's almost just as well to assume there will be no further updates, and even then it's not certain what those updates are necessarily going to be. Adding various TODO statements on what will be changed could be helpful to communicate this and to help guide future changes.